solutionbox/structured_data/datalab_solutions/structured_data/predict/predict.py

```diff
@@ -0,0 +1,434 @@
+                      help="""
+      The output results. 
+        raw_json: produces a newline file where each line is json. No 
+            post processing is performed and the output matches what the trained
+            model produces.
+        csv: produces a csv file without a header row and a header csv file.
+            For classification problems, the vector of probabalities for each 
+            target class is split into individual csv columns.""")
+
+  args, _ = parser.parse_known_args(args=argv[1:])
+
+  if args.cloud:
+    if not args.project_id:
+      raise ValueError('--project_id needed with --cloud')
+    if not args.trained_model_dir.startswith('gs://'):
+      raise ValueError('trained_model_dir needs to be a GCS path,')
+    if not args.output_dir.startswith('gs://'):
+      raise ValueError('output_dir needs to be a GCS path.')
+    if not args.predict_data.startswith('gs://'):
+      raise ValueError('predict_data needs to be a GCS path.')
+
+
+  return args
+
+
+class FixMissingTarget(beam.DoFn):
+  """A DoFn to fix missing target columns."""
+
+  def __init__(self, trained_model_dir):
+    """Reads the schema file and extracted the expected number of columns.
+
+    Args:
+      trained_model_dir: path to model.
+
+    Raises:
+      ValueError: if schema.json not found in trained_model_dir
+    """
+    from tensorflow.python.lib.io import file_io
+    import json
+    import os
+
+    schema_path = os.path.join(trained_model_dir, 'schema.json')
+    if not file_io.file_exists(schema_path):
+      raise ValueError('schema.json missing from %s' % schema_path)
+    schema = json.loads(file_io.read_file_to_string(schema_path))
+    self._num_expected_columns = len(schema)
+
+  def process(self, context):
+    import logging
+    import apache_beam as beam
+
+    num_columns = len(context.element.split(','))
+    if num_columns == self._num_expected_columns:
+      yield context.element
+    elif num_columns + 1 == self._num_expected_columns:
+      yield ',' + context.element
+    else:
+      logging.error('Got an unexpected number of columns from [%s].' %
+                    context.element)
+      yield beam.pvalue.SideOutputValue('errors',
+                                        ('bad columns', context.element))
+
+
+class EmitAsBatchDoFn(beam.DoFn):
+  """A DoFn that buffers the records and emits them batch by batch."""
+
+  def __init__(self, batch_size):
+    """Constructor of EmitAsBatchDoFn beam.DoFn class.
+
+    Args:
+      batch_size: the max size we want to buffer the records before emitting.
+    """
+    self._batch_size = batch_size
+    self._cached = []
+
+  def process(self, context):
+    self._cached.append(context.element)
+    if len(self._cached) >= self._batch_size:
+      emit = self._cached
+      self._cached = []
+      yield emit
+
+  def finish_bundle(self, context):
+    if len(self._cached) > 0:  # pylint: disable=g-explicit-length-test
+      yield self._cached
```

Comment:
Assuming this yields last batch, does it need to be padded so the batch size remains the same?